aten/src/ATen/native/utils/ParamsHash.h

```diff
@@ -38,5 +38,59 @@ struct ParamsEqual {
   }
 };
 
+// Provide explicit byte-for-byte constructors to avoid uwittingly leaving
+// padding bytes unitialized (e.g., when passing Params by value)
+template <typename T>
+struct ParamsWrapper {
+  T pod;
+  static_assert(std::is_standard_layout_v<T>, "ParamsWrapper cannot wrap non-POD data");
+
+  ParamsWrapper() {
+    memset(&(this->pod), 0, sizeof(T));
+  }
+
+  ParamsWrapper(const ParamsWrapper<T> &other) {
+    memcpy(&(this->pod), &(other.pod), sizeof(T));
+  }
+
+  ParamsWrapper(ParamsWrapper<T> &&other) {
+    memcpy(&(this->pod), &(other.pod), sizeof(T));
+  }
+
+  ParamsWrapper& operator=(const ParamsWrapper<T> &other) {
+    memcpy(&(this->pod), &(other.pod), sizeof(T));
+    return *this;
+  }
+
+  ParamsWrapper& operator=(ParamsWrapper<T> &&other) {
+    memcpy(&(this->pod), &(other.pod), sizeof(T));
+    return *this;
+  }
+
+  inline friend bool operator==(const ParamsWrapper<T> &lhs, const ParamsWrapper<T> &rhs) {
```

Comment:
It seems that this is needed because C++ doesn't like providing a default comparator (because it can't see into the type of `T` ahead of time)?

Reply:
Would it work if we add `operator==(...) = default` to `CacheKey` and `CacheKeyFused`? If the `=default` does not work (because it is C++20), can we implement a manual item-by-item comparison for these two classes? 

Considering that we are doing a separate hash struct, I think we can also implement an item-by-item hash using utilities from https://github.com/pytorch/pytorch/blob/main/c10/util/hash.h? Something like:
```C++
template <typename Params>
struct ParamsHash;

template <>
struct ParamHash <CacheKey> {
...
};

template <>
struct ParamHash <CacheKeyFused> {
...
};
```

I think implementing a manual item-by-item hash should be simpler than the current wrapper approach.

Reply:
I'm not sure I understand how the item-by-item hash would work. Would it still be generic across all key types, or would each Key type with its fields need to be manually added?

Right now `ParamsHash` is used to hash `ConvolutionParams`, a V8 Convolution params variant, `cuFFTParams`, quantized   binary cuDNN v8 params, quantized cuDNN v8 conv params, and quantized cuDNN v8 linear params. Given the number of key types that depend on this, I'd hopefully like to avoid writing/maintaining custom hash for each of them if that makes sense.

Reply:
e.g., my understanding is that we would have a lot of duplicated code that looks like:
```
// custom specialization of std::hash can be injected in namespace std
template<>
struct std::hash<S>
{
    [std::size_t](https://en.cppreference.com/w/cpp/types/size_t) operator()(S const& s) const noexcept
    {
        [std::size_t](https://en.cppreference.com/w/cpp/types/size_t) h1 = std::hash<[std::string](https://en.cppreference.com/w/cpp/string/basic_string)>{}(s.first_name);
        [std::size_t](https://en.cppreference.com/w/cpp/types/size_t) h2 = std::hash<[std::string](https://en.cppreference.com/w/cpp/string/basic_string)>{}(s.last_name);
        return h1 ^ (h2 << 1); // or use boost::hash_combine
    }
};
```

Reply:
Oh, I see. I thought it is just `CacheKey` and `CacheKeyFused`.